### ISSUE TITLE
Examples of R embedded in Java-Processing

### ISSUE THREAD

**Issue by jeremydouglass** [gaocegege/Processing.R#71]
This came up in a [discussion on the Renjin listserv](https://groups.google.com/forum/#!topic/renjin-dev/rrLM4wo3by8): an example of using renjin inside a Processing(Java) sketch, rather than the other way around (using Renjin to power a Processing mode).

-  https://gist.github.com/mjkallen/8299891

It isn't directly relevant to Processing.R -- but someday it might actually be interesting to include an example of that in the Processing.R example set. Install the R mode in PDE, switch to a Java sketch, then import renjin from the R mode. Not the core goal of this project, but potentially of interest -- and it expands the options for community engagement!

**Comment by gaocegege** [comment]
Yes, I had the idea to make java and R work together. I am not sure in what form it may take, but it really interesting 😄

**Comment by gaocegege** [comment]
If we could hide renjin from the code, it will be awesome:

```java
void setup() {
  double x = 19;
  ```R
  y <- rpois(100,lambda=3)
  ```
  for (int i = 0; i < y.size(); i++) {
    line(x, y[i])
  }
}
```

**Comment by jeremydouglass** [comment]
Not *impossible* to implement -- Jupyter notebooks, Beaker, and other notebook-paradigm multi-language systems implement mixing of R into scripting languages like Python or Javascript use an approach that is something like your example above, using "magic commands".

**However**, I think this would be way, way too much to take on for GSOC.

A more reasonable thing would be to create an example that simply re-implements the  basic mjkallen example, only pointing to the Processing.R mode's local jar (which is the same "jar-with-dependencies" as the example). Once that worked, we might choose to develop more examples in the future, or not.

For a future post-GSOC project, another way of embedding R into Java using the Python.R mode might be something like the equivalent of the way that PShaders work in Processing. 

This may already exist


Looks very good! I'll email you about a discussion date.

**Comment by gaocegege** [comment]
Jupyter doesn't allow the interaction between Java and R, IMO. For example, using variable in Java as a parameter in a function in R code.

And mjkallen's example is more like Java code, not R code. If we support Java in Processing.R, it is a little weird, IMO. 

The post-GSoC idea seems cool and feasible, I will have a look on pshader.

**Comment by jeremydouglass** [comment]
That's right, I was giving Jupyter magicks as an example of mixing R and Python, not R and Java. I believe that Beaker notebooks allow mixing R and Java. Still, those are huge projects.

I'm not suggesting that Processing.R should change. I'm pointing out that installing Processing.R installs renjin. Processing.R *could* come with a bunch of R mode examples AND a single Java mode "embedded R" example. While in Java mode, that Embedded R example would show how to call the locally installed renjin as per mjkallen's sketch.

**Comment by gaocegege** [comment]
That is a good idea 🤔

**Comment by jeremydouglass** [comment]
I created a test Processing(Java) sketch that implements this idea. It works, but it requires that the sketch `/code` folder contain a renjin jar file:

> /code/renjin-script-engine-0.8.2413-jar-with-dependencies.jar

It would be nice if the sketch could simply import the part of renjin that it needs directly from the Processing.R mode! Then installing Processing.R would also simultaneously enable basic R-in-Java mode. However, I think that the `modes` directory is not in the Java classpath for `import`. I am not sure if there could be a workaround.

```
// https://gist.github.com/mjkallen/8299891
import grafica.*;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import org.renjin.sexp.*;

void setup() {
  size(500, 350);
  background(150);
  // Prepare the points for the plot
  int nPoints = 100;

  GPointsArray points = new GPointsArray(nPoints);
  try {
    points = Random(nPoints);
  } catch (Exception e) {
    e.printStackTrace();
  }

  // Create a new plot and set its position on the screen
  GPlot plot = new GPlot(this);
  plot.setPos(25, 25);

  // Set the plot title and the axis labels
  plot.setPoints(points);
  plot.getXAxis().setAxisLabelText("x axis");
  plot.getYAxis().setAxisLabelText("y axis");
  plot.setTitleText("A very simple example");

  // Draw it!
  plot.defaultDraw();
}

GPointsArray Random(int nPoints) throws Exception {
  GPointsArray points = new GPointsArray(nPoints);
  // create a script engine manager
  ScriptEngineManager factory = new ScriptEngineManager();
  // create a Renjin engine
  ScriptEngine engine = factory.getEngineByName("Renjin");
  // evaluate R code from String
  DoubleVector res = (DoubleVector)engine.eval("runif(100)");
  for(int i = 0; i < nPoints; i++){
    points.add(i, (float)(10*res.getElementAsDouble(i)));
  }
  return points;
}
```
